### PR TITLE
BUG: partially-inferring pydatetime objects

### LINKED ISSUE #40111: API/BUG: partially-inferring datetimes
from `test_mixed_offsets_with_native_datetime_raises`

```
vals = [
    "nan",
    pd.Timestamp("1990-01-01"),
    "2015-03-14T16:15:14.123-08:00",
    "2019-03-04T21:56:32.620-07:00",
    None,
]
ser = pd.Series(vals)
```

As expected, this comes back as object-dtype, but it also does some casting:

```
>>> ser
0                                 NaT
1                 1990-01-01 00:00:00
2    2015-03-14 16:15:14.123000-08:00
3    2019-03-04 21:56:32.620000-07:00
4                                None
dtype: object
```

`ser[0]` is a _str_ "NaT" and `ser[[2, 3]]` are stdlib datetime objects.  ser[1] and ser[4] are unchanged.

This casting is done inside a call to `to_datetime`, but has the surprising effect of not being idempotent:

```
>>> pd.to_datetime(ser)
ValueError: Tz-aware datetime.datetime cannot be converted to datetime64 unless utc=True
```

I'd expect `ser` to be constructed without doing any casting.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] xref #40111
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/tools/test_to_datetime.py [hunk 1]
```diff
@@ -1123,17 +1123,32 @@ def test_iso8601_strings_mixed_offsets_with_naive(self):
...
-                "2015-03-14T16:15:14.123-08:00",
-                "2019-03-04T21:56:32.620-07:00",
+                Timestamp("2015-03-14T16:15:14.123-08:00").to_pydatetime(),
+                Timestamp("2019-03-04T21:56:32.620-07:00").to_pydatetime(),
                 None,
-            ]
+            ],
+            dtype=object,
         )
+        tm.assert_series_equal(mixed, expected)
```
Comment (simonjayhawkins) [inline pandas/tests/tools/test_to_datetime.py:1148]: does this PR fully close the issue? The issue refers to this as not being idempotent.

File: pandas/tests/tools/test_to_datetime.py [hunk 2]
```diff
@@ -1123,17 +1123,32 @@ def test_iso8601_strings_mixed_offsets_with_naive(self):
...
+
         with pytest.raises(ValueError, match="Tz-aware datetime.datetime"):
-            to_datetime(ser)
+            to_datetime(mixed)
```
Comment (simonjayhawkins) [inline pandas/tests/tools/test_to_datetime.py:1151]: this second call to `to_datetime` should be a no-op?

Comment (jbrockmendel) [reply pandas/tests/tools/test_to_datetime.py:1151]: likely; treating that as separate

File: pandas/tests/tools/test_to_datetime.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/tools/test_to_datetime.py:1148]: can keep it open with a comment that the idempotency has not been addressed

Comment (simonjayhawkins) [reply pandas/tests/tools/test_to_datetime.py:1148]: sgtm

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
cc @mroeschke if any comments

**Review by mroeschke — APPROVED** [review]

**Pull request merged**